### PR TITLE
feat(WhisperManagerPlugin): Add new Whisper Manager plugin

### PR DESCRIPTION
**Pull request by LagunaElectric**
Check it out, it's a whisper manager!

### Features:

- Each player gets their own tab with unread indicators.
- System and mod messages get their own tabs, too!
- Chat history is read in on page load, so this will work with your existing conversations even if it's your first time 😊

### CODE REVIEW

**Comment by LagunaElectric** [comment]
Unread indicator shows up even if you're the one who sent the message. Cheers to Houston for pointing this out. Will fix soon (~1hr).

**Comment by LagunaElectric** [comment]
last commit was no good. Figuring out a better way.

**Comment by LagunaElectric** [comment]
Fixed it :D

File: package-lock.json
Comment (Ishadijcks) [inline package-lock.json:1]: Can you revert these changes to the lockfile?

File: src/MooLite/core/Game.ts [hunk 1]
```diff
@@ -30,6 +33,7 @@ export class Game {
 
     lastPing: Date | null = null;
     lastPong: Date | null = null;
+    chatWhisperHistory: ChatMessage[] = [];
```
Comment (Ishadijcks) [inline src/MooLite/core/Game.ts:36]: This makes more sense as part of the `Chat` class

File: src/MooLite/core/character/Character.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export interface Character {
```
Comment (Ishadijcks) [inline src/MooLite/core/character/Character.ts:1]: I would prefer for `Character` to be a class so we can add helper methods to it if needed.

It could implement an interface with this content called `CharacterDetail` to ensure it has the right attributes

That shouldn't affect code relying on `game.character`

File: src/MooLite/core/server/messages/InitCharacterInfoParser.ts [hunk 3]
```diff
@@ -14,18 +14,22 @@ import { ActionTypeHrid } from "src/MooLite/core/actions/ActionTypeHrid";
...
+import { Character } from "src/MooLite/core/character/Character";
+import { ChatMessage } from "src/MooLite/core/chat/ChatMessage";
 
 export interface InitCharacterInfoMessage extends ServerMessage {
     type: ServerMessageType.InitCharacterInfo;
 
     actionTypeDrinkSlotsMap: Record<ActionTypeHrid, CharacterConsumable[]>;
     actionTypeFoodSlotsMap: Record<ActionTypeHrid, CharacterConsumable[]>;
     abilityCombatTriggersMap: Record<AbilityHrid, CombatTrigger>;
+    character: Character;
```
Comment (Ishadijcks) [inline src/MooLite/core/server/messages/InitCharacterInfoParser.ts:26]: This would then be `character: CharacterDetail`

File: src/MooLite/plugins/WhisperManager/DevConstants.ts [hunk 4]
```diff
@@ -0,0 +1,565 @@
+import { ChatChannelTypeHrid } from "src/MooLite/core/chat/ChatChannelTypeHrid";
+import { ConversationMessage } from "src/MooLite/plugins/WhisperManager/ConversationMessage";
+
+export const contributorNames: string[] = ["LagunaE", "Sleutelkind"];
+export const betaTesterNames: string[] = ["JBQ", "knighterd", "Thessaloniki", "Houston"];
+export const helperNames: string[] = ["Champ", "ML", "Maxion", "hoador"];
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/DevConstants.ts:6]: I actually don't like some people having a special name.

Cheze is working on a chat icon for community contributors, and I'm sure the relevant people will be accredited that way. It's not really up to us

File: src/MooLite/plugins/WhisperManager/DevConstants.ts [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+
+export const conversationData: Record<string, ConversationMessage[]> = {
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/DevConstants.ts:8]: This Dev data can either be removed or rewritten into a test case

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        icon: "💬",
+        pluginName: this.name,
+        componentName: "WhisperManagerPluginDisplay",
+        component: markRaw(WhisperManagerPluginDisplay),
+        width: "w-80",
+    };
+
+    initialize(game: Game): void {
+        super.initialize(game);
+        console.log("WhisperManagerPlugin initialized");
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:50]: Remove debug print, I think you can remove this entire `initialize()`

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        conversations = this.hideModMessages ? (({ Mods, ...o }) => o)(conversations) : conversations;
+
+        return conversations;
+    }
+
+    public get game(): Game {
+        return this._game;
+    }
+
+    public populateDevConversations(): void {
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:86]: Also remove/move to test

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+            </h1>
+            <button
+                v-if="isDev"
+                class="hover:scale-125 transition-transform aspect-square h-7 text-lg rounded-full text-gray-400 bg-gray-800 shadow-md hover:bg-gray-700"
+                @click="console.log('info', conversations)"
+            >
+                ℹ
+            </button>
+        </div>
+        <Accordion v-if="isDev">
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:102]: Also remove

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                class="bg-gray-800 font-semibold p-1.5 rounded-[4px] text-gray-400 hover:text-gray-300 hover:bg-gray-700"
+                @click="unhideAllConversations()"
+            >
+                Unhide All
+            </button>
+            <input
+                v-if="Object.keys(conversations).filter((name) => !conversations[name].hidden).length > 4"
+                v-model="searchText"
+                type="text"
+                placeholder="Search tabs..."
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:138]: I would change this to something like "search users" to avoid giving the impression you can search form message content

File: src/components/chat/ChatElement.vue [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+</script>
+
+<template>
+    <div
+        class="flex flex-row items-end space-x-1 w-full"
+        :class="{
+            'justify-end': !message.isInbound,
+        }"
+    >
+        <svg
```
Comment (Ishadijcks) [inline src/components/chat/ChatElement.vue:17]: Please refactor svgs to custom components so they are easier to understand semantically

Comment (Ishadijcks) [reply src/components/chat/ChatElement.vue:17]: And we can reuse them later :)

File: src/components/atoms/Accordion.vue [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+<template>
+    <div>
+        <button
+            @click="toggleAccordion()"
+            class="flex items-center justify-between w-full p-2 text-sm font-semibold text-left text-gray-400 bg-gray-800 rounded-[4px] hover:bg-gray-700 focus:outline-none focus:shadow-outline"
+            :aria-expanded="isOpen"
+        >
+            <slot name="title" />
+
+            <svg
```
Comment (Ishadijcks) [inline src/components/atoms/Accordion.vue:25]: Please refactor svgs to custom components

File: src/components/atoms/Accordion.vue [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        },
+    },
+};
+</script>
+
+<template>
+    <div>
+        <button
+            @click="toggleAccordion()"
+            class="flex items-center justify-between w-full p-2 text-sm font-semibold text-left text-gray-400 bg-gray-800 rounded-[4px] hover:bg-gray-700 focus:outline-none focus:shadow-outline"
```
Comment (Ishadijcks) [inline src/components/atoms/Accordion.vue:20]: Don't use any colors not from the custom theme to avoid them clashing if they change

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+                            c6.074-6.075,6.074-15.909,0-21.986L285.08,230.397z"
+                        />
+                    </svg>
+                </div>
+            </div>
+            <MooDivider style="margin-top: 0" />
+            <div class="relative h-full">
+                <ChatBox :messages="activeConversation?.messages ?? []" class="absolute inset-0 flex-1" />
+            </div>
+            <input
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:206]: Remove this input when on the `System` tab

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+    // 'change' instead of 'input', see https://github.com/facebook/react/issues/11488#issuecomment-381590324
+    el.dispatchEvent(new Event("change", { bubbles: true }));
+}
+
+const sendWhisper = (recipient: string, message: string) => {
+    const chatInput = document.getElementsByClassName("Chat_chatInput__16dhX")[0] as HTMLInputElement;
+    const sendButton = Array.from(document.getElementsByClassName("Button_button__1Fe9z")).find(
+        (button) => button.textContent === "Send"
+    ) as HTMLButtonElement;
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:56]: Please add some TODO's here to refactor once UI reflection is in place

File: src/components/atoms/Accordion.vue
Comment (Ishadijcks) [inline src/components/atoms/Accordion.vue:1]: Nice, we can reuse that!

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+            <input
+                v-if="Object.keys(conversations).filter((name) => !conversations[name].hidden).length > 4"
+                v-model="searchText"
+                type="text"
+                placeholder="Search tabs..."
+                class="p-1 bg-gray-800 rounded-[4px] my-2 text-gray-400 w-full"
+            />
+            <div class="flex flex-row space-x-0.5 overflow-y-visible overflow-x-auto">
+                <div
+                    v-for="[user, conversation] in Object.entries(conversations)
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:143]: Put this large logic sequence into a computed object to improve readability

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+        <div v-else class="flex flex-col space-y-1.5 h-full max-h-full mb-1.5">
+            <button
+                v-if="Object.keys(conversations).filter((name) => conversations[name].hidden).length"
+                class="bg-gray-800 font-semibold p-1.5 rounded-[4px] text-gray-400 hover:text-gray-300 hover:bg-gray-700"
+                @click="unhideAllConversations()"
+            >
+                Unhide All
+            </button>
+            <input
+                v-if="Object.keys(conversations).filter((name) => !conversations[name].hidden).length > 4"
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:135]: Put this logic sequence into a computed object to improve readability

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+        if (!this._conversations[name]) return;
+        this._activeConversationName = name;
+        if (this._conversations[name].unread) {
+            this._conversations[name].toggleUnread();
+        }
+    }
+
+    onChatMessage(message: ChatMessage): void {
+        const { channel, isModMessage, isSystemMessage } = message;
+        const theseAreTheDroidsWereLookingFor =
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:109]: Funny, but please rename to something sensible 😄

**Review by Ishadijcks — CHANGES_REQUESTED** [review]
Thanks for your custom plugin, it works really well!

Major points of feedback
- Don't use custom colors not from the MWI theme. If Cheze ever changes the color scheme, these won't change with them and things might clash
- Don't inline svgs, but make them into custom components. This improves code readability as I can scan the component name instead of a path. And we can reuse them later. You can put them all into `components/icons`
- Avoid making any computations in the Vue template. It is much easier to read if all filtering/mapping/sorting has been done in computables beforehand

I'll do a more thorough review once these points are fixed

Thanks for your contribution and let's work together to get this merged ASAP 💪

File: package-lock.json
Comment (AS-Laguna) [reply package-lock.json:1]: How can I do that for a single file? This happened when I did `npm install` the first time, btw.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts (hunk 17 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:109]: Aw c'mon, let me keep this one :D I would just rename it to something like `shouldHandleMessage`. I'll change if you insist but I like it enough to leave this comment. Worth a shot.

File: src/MooLite/plugins/WhisperManager/DevConstants.ts (hunk 4 above)
Comment (AS-Laguna) [reply src/MooLite/plugins/WhisperManager/DevConstants.ts:6]: I think you misunderstood. This is for people who helped, tested, or contributed to this plugin. It's a small thanks to them :D I'm happy to remove you from the list if you like.

File: package-lock.json
Comment (AS-Laguna) [reply package-lock.json:1]: I'll figure this out.

Comment (Ishadijcks) [reply package-lock.json:1]: There is some weirdness with it in dev mode, Maybe your changes is actually how it should be, but I prefer that to be a separate PR #99

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts (hunk 6 above)
Comment (AS-Laguna) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:50]: Yep.

File: src/MooLite/plugins/WhisperManager/DevConstants.ts (hunk 5 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/DevConstants.ts:8]: I didn't see that any test framework was being used here. Do we want to include one?

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts (hunk 7 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:81]: pending [this question](https://github.com/Ishadijcks/MooLite/pull/96#discussion_r1243000427).

File: src/MooLite/plugins/WhisperManager/DevConstants.ts (hunk 4 above)
Comment (Ishadijcks) [reply src/MooLite/plugins/WhisperManager/DevConstants.ts:6]: I understand, and I don't think that is something we should highlight in the plugin itself.

I'm all for adding an "author/maintainer" attribute to each plugin, but not change the behaviour of plugins based on who we believe deserve recognition

Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/DevConstants.ts:6]: This doesn't change the behavior at all. It simply determines the color of their name on the tabs. It's fairly common for beta testers and the like to be distinguished somehow in the final product. Look at MWI itself with the beta bunnies.

I disagree. The people listed here were instrumental in helping me pump this out so fast. I spoke with them about the colors they like, fiddled with it to get the contrast right, and checked that they like it.

Can we just restrict this feature to it's current implementation instead of removing it? I don't have to use these anywhere else, but I think this use case is fine.

File: src/components/atoms/Accordion.vue (hunk 12 above)
Comment (LagunaElectric) [reply src/components/atoms/Accordion.vue:22]: I'll need to do this in a few places in the app. I'll do a pass after work today and migrate everything over to the theme colors!

**Comment by LagunaElectric** [comment]
Current state of the UI:

image

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+        if (isNonStandardMessage) {
+            this._addToConversations(message, true);
+            return;
+        }
+
+        this._messages.push(message);
+    }
+
+    _addToConversations(message: ChatMessage, markUnread?: boolean): void {
+        const conversationMessage = message as ConversationMessage;
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:108]: You can do something like this to avoid casting and possible future bugs!

```ts
        const conversationMessage: ConversationMessage = {
            ...message,
            isInbound: message.receiverName === this._game.character.name,
        };
```

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+
+const props = defineProps<{
+    plugin: WhisperManagerPlugin;
+}>();
+
+const chatMessageText = ref("");
+const chatMessageInput = ref(null);
+const searchText = ref("");
+const characterName = props.plugin.game.character.name;
+const isDev = characterName === "LagunaE";
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:18]: Please remove dev stuff

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+                    'bg-clip-text hover:text-transparent bg-gradient-to-br from-orange-400 to-scarlet-400 to-90% via-orange-600 via-30%':
+                        betaTesterNames.includes(characterName),
+                    'bg-clip-text hover:text-transparent bg-gradient-to-br from-burble-300 to-burble-400 to-90%':
+                        helperNames.includes(characterName),
+                }"
+            >
+                Whisper Manager
+            </h1>
+            <button
+                v-if="isDev"
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:123]: Please remove dev stuff

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+                class="hover:scale-125 transition-transform aspect-square h-7 text-lg rounded-full text-dark-mode bg-midnight-500 shadow-md hover:bg-midnight-300"
+                @click="console.log('info', conversations)"
+            >
+                ℹ
+            </button>
+        </div>
+        <div
+            v-if="!Object.keys(conversations).length"
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:131]: Replace with computed `!hasConversations`

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+                    </span>
+                    <CrossIcon
+                        v-if="activeConversationName === user"
+                        class="h-5 aspect-square fill-dark-mode stroke-3 rounded-full p-1 transition-colors hover:fill-progress hover:bg-neutral-200"
+                        @click="conversation.hide()"
+                    />
+                </div>
+            </div>
+            <!-- Have to add this as a style tag or it gets overridden. -->
+            <MooDivider style="margin-top: 0" />
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:183]: Does the `mt-0` class not work?

File: src/components/chat/ChatBox.vue [hunk 23]
```diff
@@ -0,0 +1,17 @@
+<script setup lang="ts">
+import { defineProps } from "vue";
+import ChatElement from "./ChatElement.vue";
```
Comment (Ishadijcks) [inline src/components/chat/ChatBox.vue:3]: Could you make this import absolute?

**Review by Ishadijcks — CHANGES_REQUESTED** [review]
Very nearly done, just a few small changes!

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+
+    public populateConversations(): void {
+        if (Object.keys(this._conversations).length > 0) return;
+        for (const message of this._game.chat.whisperHistory) {
+            this._addToConversations(message);
+        }
+    }
+
+    public setActiveConversation(name: string): void {
+        if (!this._conversations[name]) return;
```
[the remark this answers is not in the document]
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:88]: ?

[the remark this answers is not in the document]
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:88]: No one line if returns?

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+import { Conversation } from "src/MooLite/plugins/WhisperManager/Conversation";
+import { PluginConfig } from "src/MooLite/core/plugins/config/PluginConfig";
+import { PluginConfigType } from "src/MooLite/core/plugins/config/PluginConfigType";
+
+export class WhisperManagerPlugin extends MooLitePlugin {
+    name: string = "Whisper Manager";
+    key = "whisper-manager";
+    description: string = "A plugin to help manage all of your conversations!";
+
+    _messages: ChatMessage[] = [];
```
[the remark this answers is not in the document]
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:18]: Nope, removed.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts (hunk 18 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:108]: Done!

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue (hunk 22 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:183]: No, it is overridden.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue (hunk 21 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:131]: Done.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue (hunk 20 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:123]: Done.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue (hunk 19 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:18]: Done.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (Ishadijcks) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:88]: Yeah, but I just noticed I haven't even setup eslint lol

It's fine for now

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue (hunk 22 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:183]: Also I plan to build a full-fledged tab widget in the design system, so this will be replaced in the future.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+
+    public get messages(): ChatMessage[] {
+        return this._messages;
+    }
+
+    public get conversations(): Record<string, Conversation> {
+        let conversations: Record<string, Conversation> = this.hideSystemMessages
+            ? (({ System, ...o }) => o)(this._conversations)
+            : this._conversations;
+        conversations = this.hideModMessages ? (({ Mods, ...o }) => o)(conversations) : conversations;
```
Comment (Ishadijcks) [inline src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:71]: I have a hard time understanding what is going on here.
Is a [.filter()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) not sufficient?

File: src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue (hunk 22 above)
Comment (Ishadijcks) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPluginDisplay.vue:183]: :+1:

File: src/components/chat/ChatBox.vue (hunk 23 above)
Comment (LagunaElectric) [reply src/components/chat/ChatBox.vue]: Done.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts (hunk 26 above)
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:66]: No, because this is an object and not an array. This is the way to omit entries from an object without using lodash.

File: src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (LagunaElectric) [reply src/MooLite/plugins/WhisperManager/WhisperManagerPlugin.ts:83]: I disagree here. I think this syntax should be used for gate checks like this. If it was anything other than a return, sure. Does this make sense? Imagine I have a couple of checks. That's 4-6 extra lines for very little benefit.

**Review by Ishadijcks — APPROVED** [review]
LGTM

I'll rework the credit system via #102, I like idea of giving fun easter eggs to the people who contributed
